src/lib/mina_base/parties.ml

```diff
@@ -158,6 +183,74 @@ module Call_forest = struct
   let accumulate_hashes_predicated xs =
     accumulate_hashes ~hash_party:Party.Predicated.digest xs
 
+  let add_callers (type party party_with_caller digest id)
+      (ps : (party, digest) t) ~(call_type : party -> Party.Call_type.t)
+      ~(add_caller : party -> id -> party_with_caller) ~(null_id : id)
+      ~(party_id : party -> id) : (party_with_caller, digest) t =
+    let rec go curr_caller parent_id ps =
+      let id_for_party p =
+        match call_type p with
+        | Delegate_call ->
```

Comment:
I think I misspoke: isn't the behaviour of a `Call` in a `Delegate_call` wrong? That is, won't the inner call resolve to the direct caller instead of its parent?

I would have expected that, if `A` `DelegateCall`s `B`, then `B` `Call`s `C`, it should appear to `C` that `A` called it, but the logic here will present `B` as the caller.

Reply:
Yeah, I think you are right. Following the semantics of ethereum's delegatecall I think delegatecall should mean "do not update the caller when we push into this node" and call should mean "update the caller when we push into this node".